File: packages/territory-rules/src/cap.ts
/** Weekly walking cap (`plan.md` "Shared Rule Semantics" item 8). */
import { RULES } from './config.js';
import type { CappedContribution, Contribution } from './types.js';

function compareStrings(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/**
 * Sum metres per `(cell, factionId, userId)` and cap each sum at
 * `WEEKLY_CAP_M_PER_PLAYER_PER_CELL` (2000 m). One row per group (`meters` is the uncapped
 * sum), sorted by cell, factionId, userId. Capture bonuses never pass through this function.
 */
export function applyWeeklyCap(contributions: readonly Contribution[]): CappedContribution[] {
  const groups = new Map<string, CappedContribution>();
  for (const c of contributions) {
    const key = `${c.cell} ${String(c.factionId)} ${c.userId}`;
    const row = groups.get(key);
    if (row) {
      row.meters += c.meters;
    } else {
      groups.set(key, {
        cell: c.cell,
        factionId: c.factionId,
        userId: c.userId,
        meters: c.meters,
        cappedMeters: 0,
      });
    }
  }
  return [...groups.values()]
    .map((r) => ({
      ...r,
      cappedMeters: Math.min(r.meters, RULES.WEEKLY_CAP_M_PER_PLAYER_PER_CELL),
    }))
    .sort(
      (a, b) =>
        compareStrings(a.cell, b.cell) ||
        a.factionId - b.factionId ||
        compareStrings(a.userId, b.userId),
    );
}

File: apps/ios/Packages/TerritoryRules/Sources/TerritoryRules/Cap.swift
import Foundation
import H3Kit

/// Sums metres per `(cell, factionId, userId)` and caps each sum at `Rules.weeklyCapMPerPlayerPerCell`
/// (plan.md item 8). Bonuses are never passed through here. Output is sorted by cell, faction, user.
public func applyWeeklyCap(_ contributions: [Contribution]) -> [CappedContribution] {
    struct Key: Hashable {
        let cell: H3Index
        let factionId: Int
        let userId: String
    }

    var sums: [Key: Double] = [:]
    for contribution in contributions {
        let key = Key(cell: contribution.cell, factionId: contribution.factionId, userId: contribution.userId)
        sums[key, default: 0] += contribution.meters
    }

    return sums
        .map { key, meters in
            CappedContribution(
                cell: key.cell,
                factionId: key.factionId,
                userId: key.userId,
                meters: meters,
                cappedMeters: min(meters, Rules.weeklyCapMPerPlayerPerCell)
            )
        }
        .sorted { lhs, rhs in
            if lhs.cell != rhs.cell { return lhs.cell < rhs.cell }
            if lhs.factionId != rhs.factionId { return lhs.factionId < rhs.factionId }
            return lhs.userId < rhs.userId
        }
}
